nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -114,42 +111,157 @@ where
     //TODO -> set maximum size of path and break into multiple statements
 }
 
-pub async fn update_asset<T>(
+pub async fn upsert_asset_with_leaf_info<T>(
     txn: &T,
     id: Vec<u8>,
-    seq: Option<u64>,
-    model: asset::ActiveModel,
+    leaf: Option<Vec<u8>>,
+    seq: Option<i64>,
+    was_decompressed: bool,
 ) -> Result<(), IngesterError>
 where
     T: ConnectionTrait + TransactionTrait,
 {
-    let update_one = if let Some(seq) = seq {
-        asset::Entity::update(model).filter(
-            Condition::all()
-                .add(asset::Column::Id.eq(id.clone()))
-                .add(asset::Column::Seq.lte(seq)),
-        )
-    } else {
-        asset::Entity::update(model).filter(asset::Column::Id.eq(id.clone()))
+    let model = asset::ActiveModel {
+        id: Set(id),
+        leaf: Set(leaf),
+        leaf_seq: Set(seq),
+        ..Default::default()
     };
 
-    match update_one.exec(txn).await {
-        Ok(_) => Ok(()),
-        Err(err) => match err {
-            DbErr::RecordNotFound(ref s) => {
-                if s.contains("None of the database rows are affected") {
-                    warn!(
-                        "Update failed. No asset found for id {}.",
-                        bs58::encode(id).into_string()
-                    );
-                    Ok(())
-                } else {
-                    Err(IngesterError::from(err))
-                }
-            }
-            _ => Err(IngesterError::from(err)),
-        },
+    let mut query = asset::Entity::insert(model)
+        .on_conflict(
+            OnConflict::column(asset::Column::Id)
+                .update_columns([asset::Column::Leaf, asset::Column::LeafSeq])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    // If we are indexing decompression we will update the leaf regardless of if we have previously
+    // indexed decompression and regardless of seq.
+    if !was_decompressed {
+        query.sql = format!(
+            "{} WHERE (NOT asset.was_decompressed) AND (excluded.leaf_seq > asset.leaf_seq OR asset.leaf_seq IS NULL)",
+            query.sql
+        );
     }
+
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn upsert_asset_with_owner_and_delegate_info<T>(
```

Comment:
Hi! Any reason why `upsert_asset_with_leaf_info` and `upsert_asset_with_owner_and_delegate_info` happens as two separate queries on the `asset` table instead of a single query?

Reply:
Yes there is a reason.

First, the general form of this PR's strategy across all the indexed data is that to maintain correctness when instructions are processed out of order:
* Any "block" of data updated by a single instruction can be indexed in any order at any time.
* Any "block" of data updated by multiple instructions need to be protected by a sequence number for that specific data, and all instructions that update that particular data must update that data's sequence number.

So specifically the reason `leaf` is separate from `owner`+`delegate` is because `redeem` and `decompress` update the `leaf` but do not update the `owner`+`delegate`.

So for example if `redeem` was indexed before a `transfer` that actually preceeded it on chain, and if they were all controlled by the same `seq`, then the `owner`+`delegate` would not be updated to their correct values.

 `owner`+`delegate` can be grouped together with one sequence number for both of them because they are always updated as a pair whenever there is a leaf schema event.

Reply:
gotcha, thanks. I understand the need to separate it for `redeem` and `decompress` but for all the cases where `upsert_asset_with_owner_and_delegate_info` is called, we update the leaf as well. We should still have separate sequence numbers for `leaf` and `owner+delegate` but it those cases where we're updating both `leaf` and `owner, delegate`, couldn't they be a part of a single tx? Wouldn't it faster and lesser load during writes?

Reply:
I had them all as separate upserts for simplicity in the code, but if I can combine them for better performance I yes I will do that.

So just so I understand, do you mean creating two SQL statements but they are included in one single executed transaction?  Something like "update leaf where seq > leaf seq", and "update owner/delegate where seq > owner_delegate_seq", both added to the same tx, probably separated by semicolon when it is actually turned into SQL?

Reply:
Yes, that's what I meant. I don't know if its possible to have a single query since each of them separate conditions on them. Let me also check if that's possible but yes, would be good if its a single transaction.

Reply:
Was talking with the team and they suggested that we merge this PR, https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/87, and https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/90 which all have this same `upsert` strategy, and then after they are mainlined we can focus on optimizations.

Reply:
sounds good. Thanks!

Reply:
Ok resolving this for now will focus on optimization after we nail correctness across all instructions!